strider/query_planner.py

```diff
@@ -253,44 +253,94 @@ async def find_valid_permutations(
     return list(filtered_ogs)
 
 
-def dfs(operation_graph, source) -> (list[str], list[str]):
+def all_eulerian_paths(operation_graph, source) -> list[list[str]]:
     """
-    Run depth first search on the graph
+    Return all Eulerian paths originating from a source node.
 
-    Returns a list of nodes and edges
-    that we traversed. Will not go back to
-    previously visited nodes or edges.
+    A Eulerian path is one that visits all edges exactly once
     """
-    path_nodes = [source]
-    path_edges = []
+    paths = []
 
-    stack = [source]
+    starting_edges = [
+        edge_id for edge_id, edge in operation_graph['edges'].items()
+        if edge['source'] == source
+    ]
 
-    # The most efficient way to do a depth first search
-    # is with an adjacency list, so we build one here
+    # Initialize stack with all edges that start at our source node
+    stack = []
+    for edge in starting_edges:
+        new_graph = copy.deepcopy(operation_graph)
+        del new_graph['edges'][edge]
+        stack.append((new_graph, [edge]))
 
+    # Recursively finish paths
     while len(stack) != 0:
         s = stack.pop()
 
-        # Build adjacency list
-        adjacent_nodes = []
-        adjacent_edges = []
+        current_graph = s[0]
+        current_path = s[1]
+
+        current_node = operation_graph['edges'][current_path[-1]]['target']
+
+        adjacent_edges = [
+            edge_id for edge_id, edge in current_graph['edges'].items()
+            if edge['source'] == current_node
+        ]
+
+        if len(adjacent_edges) == 0:
+            # Done
+            paths.append(current_path)
```

Comment:
Is this guaranteed to be an Eulerian path? This seems to check only that there's nowhere else we can go from here, not that there are no edges left to be traversed.